azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationDeclarationBuilder.kt

```
@@ -0,0 +1,33 @@
+package com.anaplan.engineering.azuki.script.generation
+
+import com.anaplan.engineering.azuki.declaration.Declaration
+import com.anaplan.engineering.azuki.declaration.DeclarationBuilder
+import com.anaplan.engineering.azuki.declaration.FeDeclarationBuilderFactory
+
+interface ScriptGenerationDeclarationBuilderFactory<E : ScriptGenerationEnvironment, D : Declaration> :
+    FeDeclarationBuilderFactory<D, ScriptGenerationDeclarationBuilder<E, D>>
+
+/**
+ * Handles building a script fragment for a particular declaration.
+ */
+abstract class ScriptGenerationDeclarationBuilder<E : ScriptGenerationEnvironment, D : Declaration>(declaration: D) :
+    DeclarationBuilder<D>(declaration) {
+
+    /**
+     * Generates the declaration script.
+     * The given generation environment may be referenced or updated with information from the declaration.
+     */
+    abstract fun getDeclarationScript(environment: E): String
+
+}
+
+/**
+ * Simplified declaration builder stub for when we don't need an environment.
+ */
+abstract class NoEnvScriptGenerationDeclarationBuilder<D : Declaration>(declaration: D) :
+    ScriptGenerationDeclarationBuilder<NoScriptGenerationEnvironment, D>(declaration) {
+
+    abstract fun getDeclarationScript(): String
+
+    override fun getDeclarationScript(environment: NoScriptGenerationEnvironment) = getDeclarationScript()
+}
```

❓ Should this take the environment as a parameter?  I've found that, when working with code that _does_ use environments (but not everywhere), I'm having to duplicate this pattern locally to do the same thing, because I don't have `E = NoScriptGenerationEnvironment`.

---

I'm afraid I don't understand what you're asking here. You will need the override because it's required to implement `ScriptGenerationDeclarationBuilder`, although it feels a bit smelly because it suggests to users that they can supply an environment, but it's going to be silently ignored.

Would this be any easier with a sealed interface instead of plain inheritance, like you did for checks? (the same consistency argument applies here, by the way)

---

I don't think my original question applies anymore, so I'll resolve this.